include/seqan3/search/fm_index/fm_index_cursor.hpp

```diff
@@ -482,9 +482,14 @@ class fm_index_cursor
                       "The alphabet types of the given text and index differ.");
         assert(index != nullptr);
 
-        size_type const loc = offset() - index->index[node.lb];
-        size_type const query_begin = loc - index->text_begin_rs.rank(loc + 1) + 1; // Substract delimiters
-        return text | views::join | views::slice(query_begin, query_begin + query_length());
+        size_type const loc = offset() - index->index[node.lb];         // Position of query in concatenated texts.
+        size_type const text_idx = index->text_begin_rs.rank(loc + 1) - 1; // Position of subtext in text vector,
+                                                                        // where the query was found
+                                                                        // = rank of Bitvector of the concatenated texts
+        size_type const query_begin = loc - index->text_begin_ss.select(text_idx + 1);
+                                                                        // Substract lengths of previous sequences
+        return text[text_idx] | views::slice(query_begin, query_begin + query_length());
+                                                                        // Take subtext, slice query out of it
```

Comment:
Not sure why you subtract above and add again later. Can an underflow happen, i.e. `text_idx = -1`? Can `index->text_begin_rs.rank(loc + 1)` be `0`? Can you add `assert(text_idx + 1 != 0)`?

Is the data layout like this: "$FIRSTSTRING#SECONDSTRING%THIRDSTRING"? So that `rank` returns always  > 0, i.e. the number of delimiters? Do you know what happens when `loc + 1` is a delimiter? For example if it is a at position `%` in my example? Do you know why we take `loc + 1`? Does this avoid to be on the wrong delimiter?

Can you rearrange the code like this:

```suggestion
        // Position of query in concatenated text.
        size_type const location = offset() - index->index[node.lb];

        // The rank represents the number of delimiters before position `location + 1`.
        // Furthermore, this number represents the i-th text this text position originated from, since each delimiter corresponds to one text in the whole concatenated text from left-to-right.
        size_type const rank = index->text_begin_rs.rank(loc + 1);
        // The sum of all text lengths prior to the i-th text is the start location of the i-th text in the concatenated sequence.
        size_type const start_location = index->text_begin_ss.select(rank);

        assert(rank > 0);
        size_type const text_id = rank - 1;
        size_type const query_begin = location - start_location;
        return text[text_id] | views::slice(query_begin, query_begin + query_length());
```

Reply:
> Not sure why you subtract above and add again later. Can an underflow happen, i.e. `text_idx = -1`? Can `index->text_begin_rs.rank(loc + 1)` be `0`? Can you add `assert(text_idx + 1 != 0)`?
> Is the data layout of the concatenated text like this?:
> 
> > "$FIRSTSTRING#SECONDSTRING%THIRDSTRING"
> 
> So that `rank` returns always > 0, i.e. the number of delimiters is strictly positive? Do you know what happens when `loc + 1` is a delimiter? In other words what would happen if `location + 1` is `%` in my example? Do you know why we take `loc + 1`? Does this avoid being on the wrong delimiter and therefore giving back the wrong index of the current text?

Yes I can add `assert(text_idx + 1 != 0)`. The deliminiters should be the same, but yes I think the text looks like this.

I have made several attempts to omit the +1 and -1 or postpone it to make it clearer, but then tests always fail because we are running above or below an index.
The point is that ```index->text_begin_rs.rank(location);``` and ```index->text_begin_rs.rank(location + 1);``` can be equal or different.

If you have a look at these two tests, I changed rank to ```index->text_begin_rs.rank(location);```:
```shell
~/build/seqan3/unit_debug$ search/fm_index_cursor/fm_index_cursor_collection_test 
...
[ RUN      ] default_traits/fm_index_cursor_collection_test/0.query
-------------------
[ACGACG,TGCGATCGA]
location 0
old rank 1
rank 0
start_location 0
text_id 0
query_begin 0
result ACG
[       OK ] default_traits/fm_index_cursor_collection_test/0.query (109 ms)
...
[ RUN      ] default_traits/fm_index_cursor_collection_test/0.incomplete_alphabet
-------------------
[ATATAT,ATATAT]
location 1
old rank 1
rank 1
start_location 7
text_id 1
query_begin 18446744073709551610
result
/seqan3/test/unit/search/fm_index_cursor/fm_index_cursor_collection_test_template.hpp:319: Failure
Value of: std::ranges::equal(it.path_label(text), "T"_dna4)
  Actual: false
Expected: true
[  FAILED  ] ...
```
I have to admit, I still don't fully understand how rank works here, but everything works with (loc + 1). Maybe @smehringer  would like to add something?

Reply:
That's fine if it works for +1. We should maybe add the `+1` directly in the location variable.

I think that both the `offset` and the `index` are zero-based and that we double subtract the zero-position in the difference, so we need to add one to get the real location.

Reply:
Ok some clarification (that I hope I understood correctly from the code):

@Irallia You comments are a bit misleading (as are the non-expressive variable names in the search code :woman_facepalming:). The variable `text_begin_ss` is the select-support-data structures of the variable `text_begin` which is a bitvector that **stores the begin positions of every text in a collection`. I think that's what confused you @marehr because you thought that the bitvector marks the position of the delimiters. Thus, The rank is always greater than 0 since the very first bit will always be a 1.  

Now to the mystery of the `+1` you talked about. You must know that the `rank at position i` in the SDSL (at least I am pretty sure) is defined as `the number of 1's up until, but not including, position i`. Which means that if there is a `1` at position `i` it is missed. Thus, the `location` is correctly computed but the rank query must inquire the rank at position one after `i` in order to resolve the correct text.

> I think that both the offset and the index are zero-based and that we double subtract the zero-position in the difference, so we need to add one to get the real location.

I am not exactly sure what you meant by this sentence. Maybe you can rephrase it in case you think there is still an issue.